### PR TITLE
Set up standard library path substitution in rust-gdb and gdbgui

### LINKED ISSUE #62945: rust-gdb doesn't set up path substitutions?
I'm trying to use `rust-gdb` on project generated by `cargo new` and looks like the `rust-src` source code path is not configured:

```
Reading symbols from target/debug/hello...
(gdb) show directories
Source directories searched: ~/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/etc:$cdir:$cwd
(gdb) r
Starting program: target/debug/hello 

Program received signal SIGSEGV, Segmentation fault.
core::ops::function::FnOnce::call_once{{vtable-shim}} () at /rustc/a7f28678bbf4e16893bb6a718e427504167a9494/src/libcore/ops/function.rs:231
231	/rustc/a7f28678bbf4e16893bb6a718e427504167a9494/src/libcore/ops/function.rs: No such file or directory.
(gdb) set substitute-path /rustc/a7f28678bbf4e16893bb6a718e427504167a9494/ ~/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/
(gdb) l
226	    #[stable(feature = "fn_once_output", since = "1.12.0")]
227	    type Output;
228	
229	    /// Performs the call operation.
230	    #[unstable(feature = "fn_traits", issue = "29625")]
231	    extern "rust-call" fn call_once(self, args: Args) -> Self::Output;
232	}
233	
234	mod impls {
235	    #[stable(feature = "rust1", since = "1.0.0")] 
```

### PR DESCRIPTION
**Pull request by lnicola**
Fixes #62945

---

Only lightly tested (in release mode, where the paths are a bit of a mess) because my `gdb` appears to crash with `internal-error: inside_main_func: Assertion 'block != nullptr' failed.` and I don't have `gdbgui`. Please review carefully my shell syntax.

There's also `rust-lldb`, but I don't know the equivalent for it.

### CODE REVIEW

File: src/etc/rust-gdb [hunk 1]
```diff
@@ -21,4 +23,6 @@ RUST_GDB="${RUST_GDB:-gdb}"
 PYTHONPATH="$PYTHONPATH:$GDB_PYTHON_MODULE_DIRECTORY" exec ${RUST_GDB} \
   --directory="$GDB_PYTHON_MODULE_DIRECTORY" \
   -iex "add-auto-load-safe-path $GDB_PYTHON_MODULE_DIRECTORY" \
+  -iex "set substitute-path /rustc/$RUSTC_COMMIT_HASH $RUSTC_SYSROOT/lib/rustlib/src/rust" \
```
Comment (Mark-Simulacrum) [inline src/etc/rust-gdb:26]: This seems reasonable(ish), but I'm wondering whether this is a problem if the destination does not exist. Maybe it's clearer to see the hash path? Not sure.

Comment (cuviper) [reply src/etc/rust-gdb:26]: IME, gdb only uses the substitution for reading the source file, but still displays the original path in backtraces and such. Even the TUI layout shows the source window labeled with the original path.

**Comment by lnicola** [comment]
Sorrry for dropping the ball on this. I can try to add a new `--print` flag, it sounds reasonable. Will that have to go through a FCP?

File: src/etc/rust-gdb (hunk 1 above)
Comment (lnicola) [reply src/etc/rust-gdb:26]: IIUC, there's no issue if a path doesn't exist, the substitutions just don't work.

**Comment by cuviper** [comment]
I do still think a `--print` flag would be a good idea, but it doesn't need to block here. A [compiler MCP](https://forge.rust-lang.org/compiler/mcp.html) should be relatively lightweight for proposing that.

@bors r+ rollup

**Pull request merged**